Lib/fontTools/colorLib/builder.py

```diff
@@ -488,234 +432,71 @@ def _tuple_safe(value):
-    ):
-        ot_paint = ot.Paint()
-        ot_paint.Format = int(ot.PaintFormat.PaintTranslate)
-        ot_paint.Paint = self.buildPaint(paint)
-        ot_paint.dx = _to_variable_f16dot16_float(dx)
-        ot_paint.dy = _to_variable_f16dot16_float(dy)
-        return ot_paint
-
-    def buildPaintRotate(
-        self,
-        paint: _PaintInput,
-        angle: _ScalarInput,
-        centerX: _ScalarInput,
-        centerY: _ScalarInput,
-    ) -> ot.Paint:
-        ot_paint = ot.Paint()
-        ot_paint.Format = int(ot.PaintFormat.PaintRotate)
-        ot_paint.Paint = self.buildPaint(paint)
-        ot_paint.angle = _to_variable_f16dot16_float(angle)
-        ot_paint.centerX = _to_variable_f16dot16_float(centerX)
-        ot_paint.centerY = _to_variable_f16dot16_float(centerY)
-        return ot_paint
-
-    def buildPaintSkew(
-        self,
-        paint: _PaintInput,
-        xSkewAngle: _ScalarInput,
-        ySkewAngle: _ScalarInput,
-        centerX: _ScalarInput,
-        centerY: _ScalarInput,
-    ) -> ot.Paint:
-        ot_paint = ot.Paint()
-        ot_paint.Format = int(ot.PaintFormat.PaintSkew)
-        ot_paint.Paint = self.buildPaint(paint)
-        ot_paint.xSkewAngle = _to_variable_f16dot16_float(xSkewAngle)
-        ot_paint.ySkewAngle = _to_variable_f16dot16_float(ySkewAngle)
-        ot_paint.centerX = _to_variable_f16dot16_float(centerX)
-        ot_paint.centerY = _to_variable_f16dot16_float(centerY)
-        return ot_paint
-
-    def buildPaintComposite(
-        self,
-        mode: _CompositeInput,
-        source: _PaintInput,
-        backdrop: _PaintInput,
-    ):
-        ot_paint = ot.Paint()
-        ot_paint.Format = int(ot.PaintFormat.PaintComposite)
-        ot_paint.SourcePaint = self.buildPaint(source)
-        ot_paint.CompositeMode = _to_composite_mode(mode)
-        ot_paint.BackdropPaint = self.buildPaint(backdrop)
-        return ot_paint
-
-    def buildColrLayers(self, paints: List[_PaintInput]) -> ot.Paint:
-        ot_paint = ot.Paint()
-        ot_paint.Format = int(ot.PaintFormat.PaintColrLayers)
-        self.slices.append(ot_paint)
-
-        paints = [
-            self.buildPaint(p)
-            for p in _build_n_ary_tree(paints, n=MAX_PAINT_COLR_LAYER_COUNT)
-        ]
+    # COLR layers is unusual in that it modifies shared state
+    # so we need a callback into an object
+    def _beforeBuildPaintColrLayers(self, dest, source):
+        paint = ot.Paint()
+        paint.Format = int(ot.PaintFormat.PaintColrLayers)
+        self.slices.append(paint)
+
+        # Sketchy gymnastics: a sequence input will have dropped it's layers
+        # into NumLayers; get it back
+        if isinstance(source.get("NumLayers", None), collections.abc.Sequence):
+            layers = source["NumLayers"]
+        else:
+            layers = source["Layers"]
+
+        layers = [self.buildPaint(l) for l in layers]
+        # TODO: should treeification come after reuse?
```

Comment:
> reuse can work with the finalized state of the layer list

Seeking reuse *changes* the state of the layer list; I would argue there is no finalized layer list until after reuse has run. Reuse can make the layer list smaller by replacing sequences with PaintColrLayers. For example, it could even reduce a list from 256 layers to 255, avoiding the need to treeify at all. 

That would lead me to think we want to do something like:

1. convert layers to paints so if you see a list after building a tree you can distinguish it from a list that was meant to convert to a table
1. search for reuse, which replaces sequences of layers we've seen before with PaintColorLayers. This reduces the number of layers, which changes the tree we want, potentially eliminating the need for treeification to do anything.
   * only now do we have a stable layer list to turn into a tree
1. build the tree on the post-reuse layer set
   * now you have a mixture of Paint and lists of Paint
1. convert all the lists of Paint into PaintColrLayers
   * since you already ran reuse on them there is no need to do it again
   * the listToColrLayers function that never gets called appears to have been meant to do this job

Does that seem reasonable?

Reply:
Updated per above suggestion; builder_test.py now fully passes. It's slow on the 255 test, probably because the reuse check scales appallingly. I'm also not convinced the registration of reusable parts (for future paints to reference) works as intended when a tree is present.